MessageHandlingException consistency

artembilan opened the pull request:
* Use `IntegrationUtils.wrapInHandlingExceptionIfNecessary()` whenever
it is possible to avoid double wrapping into the `MessageHandlingException`
* Some code polishing for affected classes

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan commented:
There are several concerns in this PR:

1. The `MethodInvokingMessageProcessor` now uses `IntegrationUtils.wrapInHandlingExceptionIfNecessary`. And this is slightly a breaking change because we have some test expecting a `MessageHandlingException`, but we get a `MessageConversionException`, which is, indeed, a `MessagingException`. So, with the `IntegrationUtils.wrapInHandlingExceptionIfNecessary` contract we are good to rethrow it as is, without wrapping to the `MessageHandlingException`. I would say this is OK, but if the target application really relies only on the `MessageHandlingException`, then it is a breaking change. On the other hand the `AbstractMessageHandler` fully relies on the `IntegrationUtils.wrapInHandlingExceptionIfNecessary` and don't wrap `MessageConversionException` if that. So, this is a bit inconsistency with all our `MessageHandlingException` producers. Therefore I suggest to re-think in favor what I've done here, but already in the next `5.2` version, since this can be a breaking change for the current users.
2. I've introduced a `@NonNullApi` into the `org.springframework.integration.handler`. Fixed as much as possible according IDEA advises, but I'd appreciate your feedback to determine what I've missed and how really proceed with that `@Nullable`.
3. I did some refactoring in the affected classes (`IntegrationUtils.wrapInHandlingExceptionIfNecessary` or `@Nullable`). Mostly it is about diamonds, Java Streams for iteration and filtering, redundant methods and properties since there are some in super class.

I'll fix test-cases according Travis after your feedback.

Thanks

garyrussell reviewed (commented):
While we have to fix the exception inconsistency, I think it would be better to push it to 5.2.

I also don't think we should introduce any more `@NonNullApi` until we've resolved all the smells for the earlier ones - my last commit reduced critical smells from ~200 to ~150 but we still have a long way to go.

That said, if you want to fix the tests, I can push this to my sonar branch to see what new smells are produced.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/core/MessageProducer.java:
```diff
@@ -29,16 +30,16 @@
 	 * @param outputChannel The output channel.
 	 */
-	void setOutputChannel(MessageChannel outputChannel);
+	void setOutputChannel(@Nullable MessageChannel outputChannel);
```
OK. That's what I wanted to hear. So, yes, looking to this one more time I agree that this one should not expect `null`, but the `outputChannel` property still can be `null`.
Does it sound like we need to have a `@NonNullFields` as well? 
Or do you mean to remove a `@NonNullApi` from this package for time being and just concentrate on the smell and don't bother ourselves about `@Nullable` for now?
Or can we keep those `@Nullable` , but still pursue a consistency for null checks?

Thanks

garyrussell replied:
Yes; I would rather solve current smells first; then add new ones 😄

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/ExpressionCommandMessageProcessor.java:
```diff
@@ -51,14 +55,14 @@ public ExpressionCommandMessageProcessor(MethodFilter methodFilter) {
-		if (methodFilter != null) {
-			MethodResolver methodResolver = new ExpressionCommandMethodResolver(methodFilter);
-			this.getEvaluationContext(false).setMethodResolvers(Collections.singletonList(methodResolver));
+			setBeanFactory(beanFactory);
```
I can make it `final` in this class.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/ExpressionCommandMessageProcessor.java:
```diff
@@ -51,14 +55,14 @@ public ExpressionCommandMessageProcessor(MethodFilter methodFilter) {
+			setBeanFactory(beanFactory);
 		}
+
+		MethodResolver methodResolver = new ExpressionCommandMethodResolver(methodFilter);
```
I can keep that `if (methodFilter != null) {`. 
Then it is not what `@NonNullApi` expects from us. Or I just don't understand it properly yet...

garyrussell replied:
You understand it ok, but legacy apps (except Kotlin) are likely not to be coded to honor `@Nullable/@NonNull`.

Only apps that are coded to comply will work properly - the compiler does no assertion - just tools like Sonar/IntelliJ/Eclipse (if so configured).

In 5.2, we can make this NonNull and add an assertion (but we can't do that in 5.1).

artembilan replied:
Right. Let's this PR to hang on for a while until `5.2` then!

Or make this `methodFilter` `@Nullable` as well and still have that `if (methodFilter != null) {`...

But again let's deffer until `5.2`!

artembilan commented:
Gary, 

thank you for feedback.
Sorry, but it confirms that I don't feels things properly yet.

Of course, I would like to help with fixing those our smell, however it seems I still need to learn more about this `@NonNullApi`.

And final question: do you agree that we are good to throwing any `MessagingException` from those fixed places instead of wrapping to the `MessageHandlingException` unconditionally? 

The origin for this work came from here though: https://stackoverflow.com/questions/53596124/spring-integration-error-is-attaching-completed-payload

garyrussell commented:
Sorry, I am not sure what you're asking - we already only wrap an ME in an MHE if the `failedMessage` is `null`.

artembilan commented:
> we already only wrap an ME in an MHE if the failedMessage is null.

That's what we do in the `AbstractMH`, but the current `MethodInvokingMessageProcessor` wraps to the `MHE` unconditionally. This what I mean as an inconsistency. Any non-POJO based processing ends up with their own `ME`, but this one produces a wrapping, always:
```
	public T processMessage(Message<?> message) {
		try {
			return this.delegate.process(message);
		}
		catch (Exception e) {
			throw new MessageHandlingException(message, e);
		}
	}
```

artembilan committed:
MessageHandlingException consistency

* Use `IntegrationUtils.wrapInHandlingExceptionIfNecessary()` whenever
it is possible to avoid double wrapping into the `MessageHandlingException`
* Some code polishing for affected classes `@Nullable`, streams, diamonds etc.

artembilan commented:
OK. I think PR is ready for review.

Thanks

garyrussell commented:
You force-pushed a single commit; did you make any of the changes I suggested?

artembilan commented:
> did you make any of the changes I suggested?

Yes:

1. Removed   `@NonNullApi` from the `org.springframework.integration.handler`.
2. Fixed `ExpressionCommandMessageProcessor` for `methodFilter` null check and made its `setBeanFactory()` as `final` since it is called from the ctor.
3. Removed `@Nullable` from the `MessageProducer.setOutputChannel()`.

Seems for me that's all from your feedback.

Thanks

artembilan commented:
Also: I've squashed with the previous commit to apply `updateCopyrights` for all the affected classes in this PR.

garyrussell commented:
Thanks; I just didn't want to spend the time to verify 😄
